README.rst
==================
Bootstrapping Salt
==================

|build|

.. contents::
    :local:

Before `Salt`_ can be used for provisioning on the desired machine, the binaries need to be
installed. Since `Salt`_ supports many different distributions and versions of operating systems,
the `Salt`_ installation process is handled by this shell script ``bootstrap-salt.sh``.  This
script runs through a series of checks to determine operating system type and version to then
install the `Salt`_ binaries using the appropriate methods. For Windows, use the
``bootstrap-salt.ps1`` script.

**NOTE**

This ``README`` file is not the absolute truth as to what the bootstrap script is capable of. For
that, please read the generated help by passing ``-h`` to the script or even better,
`read the source`_.

Also, to secure your Salt installation, check out these instructions for `hardening salt`_.

Bootstrap
=========

In every two-step installation example, you would be well-served to **verify against the SHA256
sum** of the downloaded ``bootstrap-salt.sh`` file.

.. _sha256sums:

The SHA256 sum of the ``bootstrap-salt.sh`` file, per release, is:

- 2022.05.19: ``e92e1df6930285cf23eda188bee3cfa3dd6c577b4fb7aa91b29213ad820199b1``
- 2022.03.15: ``8f65952c3435f441e7f793941d5162d3ec2033a9ef82722ff1da67a2ef860a2f``
- 2021.09.17: ``090d652cd6290debce0e3a4eded65086a4272e69446e711eb26f87160593b6a2``
- 2021.09.14: ``30fdcba972f449630b4f13492cb5525e69e08fa2cdb66a6dc78f1536ad279e52``
- 2021.08.19: ``ee40a9d8d057cce88a288fc1cb94b1d31408a61d262db6f77b34ad63d66f0806``
- 2021.06.23: ``35b397dd0a50f832af453c17f138fd29e3692e492d7f463c404a57e1fac10665``
- 2021.03.02: ``91baa0073308f1be20c7be65238ef67e5733c75285314b302a5b2456e73a0758``
- 2020.10.20: ``b47bfc8d63cccf22eb4cd94491d30cc1d571e184be25a5be7f775e7f2daaf6e2``
- 2020.10.19: ``f6c3e2c52f98d115809044b09062219369957caf30228b594033f0543e202c52``
- 2020.06.23: ``1d07db867c195c864d0ae70664524f2099cc9a46872953293c67c3f239d4f4f5``
- 2020.05.28: ``6b3ea15c78f01060ab12fc01c0bb18480eaf36858c7ba188b200c0fb11aac173``
- 2020.02.24: ``efc46700aca78b8e51d7af9b06293f52ad495f3a8179c6bfb21a8c97ee41f1b7``
- 2020.02.04: ``ce877651b4938e3480f76b1629f582437f6ca8b73d7199fdb9e905e86fe85b34``
- 2020.01.29: ``e9afdfa877998c1c7f0e141a6728b33d0d24348e197aab2b9bde4fe6bc6db1b2``
- 2020.01.21: ``53299aa0dfbf7ab381f3856bb7babfc04a1d6525be11db0b9466277b1e4d0c1a``
- 2019.11.04: ``905924fccd4ebf168d19ba598bf10af53efe02302b792aeb15433e73fd3ad1d2``
- 2019.10.03: ``34f196f06d586ce9e1b9907660ea6e67caf57abcecfea66e0343697e3fd0d17d``
- 2019.05.20: ``46fb5e4b7815efafd69fd703f033fe86e7b584b6770f7e0b936995bcae1cedd8``
- 2019.02.27: ``23728e4b5e54f564062070e3be53c5602b55c24c9a76671968abbf3d609258cb``
- 2019.01.08: ``ab7f29b75711da4bb79aff98d46654f910d569ebe3e908753a3c5119017bb163``
- 2018.08.15: ``6d414a39439a7335af1b78203f9d37e11c972b3c49c519742c6405e2944c6c4b``
- 2018.08.13: ``98284bdc2b5ebaeb619b22090374e42a68e8fdefe6bff1e73bd1760db4407ed0``
- 2018.04.25: ``e2e3397d6642ba6462174b4723f1b30d04229b75efc099a553e15ea727877dfb``
- 2017.12.13: ``c127b3aa4a8422f6b81f5b4a40d31d13cec97bf3a39bca9c11a28f24910a6895``
- 2017.08.17: ``909b4d35696b9867b34b22ef4b60edbc5a0e9f8d1ed8d05f922acb79a02e46e3``
- 2017.05.24: ``8c42c2e5ad3d4384ddc557da5c214ba3e40c056ca1b758d14a392c1364650e89``

If you're looking for a *one-liner* to install Salt, please scroll to the bottom and use the
instructions for `Installing via an Insecure One-Liner`_.

There are also .sha256 files for verifying against in the repo for the stable branch.  You can also
get the correct sha256 sum for the stable release from https://bootstrap.saltproject.io/sha256 and
https://winbootstrap.saltproject.io/sha256

Contributing
------------

The Salt Bootstrap project is open and encouraging to code contributions. Please review the
`Contributing Guidelines`_ for information on filing issues, fixing bugs, and submitting features.

The `Contributing Guidelines`_ also contain information about the Bootstrap release cadence and
process.

Examples
--------

To view the latest options and descriptions for ``salt-bootstrap``, use ``-h`` and the terminal:

.. code:: console

  ./salt-bootstrap.sh -h

  Usage :  bootstrap-salt.sh [options] <install-type> [install-type-args]

  Installation types:
    - stable              Install latest stable release. This is the default
                          install type
    - stable [branch]     Install latest version on a branch. Only supported
                          for packages available at repo.saltproject.io
    - stable [version]    Install a specific version. Only supported for
                          packages available at repo.saltproject.io
                          To pin a 3xxx minor version, specify it as 3xxx.0
    - testing             RHEL-family specific: configure EPEL testing repo
    - git                 Install from the head of the master branch
    - git [ref]           Install from any git ref (such as a branch, tag, or
                          commit)

  Examples:
    - bootstrap-salt.sh
    - bootstrap-salt.sh stable
    - bootstrap-salt.sh stable 3004.1
    - bootstrap-salt.sh stable v3003.4
    - bootstrap-salt.sh testing
    - bootstrap-salt.sh git
    - bootstrap-salt.sh git 3004.1
    - bootstrap-salt.sh git v3003.4
    - bootstrap-salt.sh git 06f249901a2e2f1ed310d58ea3921a129f214358

  Options:
    -h  Display this message
    -v  Display script version
    -n  No colours
    -D  Show debug output
    -c  Temporary configuration directory
    -g  Salt Git repository URL. Default: https://github.com/saltstack/salt.git
    -w  Install packages from downstream package repository rather than
        upstream, saltstack package repository. This is currently only
        implemented for SUSE.
    -k  Temporary directory holding the minion keys which will pre-seed
        the master.
    -s  Sleep time used when waiting for daemons to start, restart and when
        checking for the services running. Default: 3
    -L  Also install salt-cloud and required python-libcloud package
    -M  Also install salt-master
    -S  Also install salt-syndic
    -N  Do not install salt-minion
    -X  Do not start daemons after installation
    -d  Disables checking if Salt services are enabled to start on system boot.
        You can also do this by touching /tmp/disable_salt_checks on the target
        host. Default: ${BS_FALSE}
    -P  Allow pip based installations. On some distributions the required salt
        packages or its dependencies are not available as a package for that
        distribution. Using this flag allows the script to use pip as a last
        resort method. NOTE: This only works for functions which actually
        implement pip based installations.
    -U  If set, fully upgrade the system prior to bootstrapping Salt
    -I  If set, allow insecure connections while downloading any files. For
        example, pass '--no-check-certificate' to 'wget' or '--insecure' to
        'curl'. On Debian and Ubuntu, using this option with -U allows obtaining
        GnuPG archive keys insecurely if distro has changed release signatures.
    -F  Allow copied files to overwrite existing (config, init.d, etc)
    -K  If set, keep the temporary files in the temporary directories specified
        with -c and -k
    -C  Only run the configuration function. Implies -F (forced overwrite).
        To overwrite Master or Syndic configs, -M or -S, respectively, must
        also be specified. Salt installation will be ommitted, but some of the
        dependencies could be installed to write configuration with -j or -J.
    -A  Pass the salt-master DNS name or IP. This will be stored under
        ${BS_SALT_ETC_DIR}/minion.d/99-master-address.conf
    -i  Pass the salt-minion id. This will be stored under
        ${BS_SALT_ETC_DIR}/minion_id
    -p  Extra-package to install while installing Salt dependencies. One package
        per -p flag. You are responsible for providing the proper package name.
    -H  Use the specified HTTP proxy for all download URLs (including https://).
        For example: http://myproxy.example.com:3128
    -b  Assume that dependencies are already installed and software sources are
        set up. If git is selected, git tree is still checked out as dependency
        step.
    -f  Force shallow cloning for git installations.
        This may result in an "n/a" in the version number.
    -l  Disable ssl checks. When passed, switches "https" calls to "http" where
        possible.
    -V  Install Salt into virtualenv
        (only available for Ubuntu based distributions)
    -a  Pip install all Python pkg dependencies for Salt. Requires -V to install
        all pip pkgs into the virtualenv.
        (Only available for Ubuntu based distributions)
    -r  Disable all repository configuration performed by this script. This
        option assumes all necessary repository configuration is already present
        on the system.
    -R  Specify a custom repository URL. Assumes the custom repository URL
        points to a repository that mirrors Salt packages located at
        repo.saltproject.io. The option passed with -R replaces the
        "repo.saltproject.io". If -R is passed, -r is also set. Currently only
        works on CentOS/RHEL and Debian based distributions.
    -J  Replace the Master config file with data passed in as a JSON string. If
        a Master config file is found, a reasonable effort will be made to save
        the file with a ".bak" extension. If used in conjunction with -C or -F,
        no ".bak" file will be created as either of those options will force
        a complete overwrite of the file.
    -j  Replace the Minion config file with data passed in as a JSON string. If
        a Minion config file is found, a reasonable effort will be made to save
        the file with a ".bak" extension. If used in conjunction with -C or -F,
        no ".bak" file will be created as either of those options will force
        a complete overwrite of the file.
    -q  Quiet salt installation from git (setup.py install -q)
    -x  Changes the Python version used to install Salt.
        For CentOS 6 git installations python2.7 is supported.
        Fedora git installation, CentOS 7, Debian 9, Ubuntu 16.04 and 18.04 support python3.
    -y  Installs a different python version on host. Currently this has only been
        tested with CentOS 6 and is considered experimental. This will install the
        ius repo on the box if disable repo is false. This must be used in conjunction
        with -x <pythonversion>.  For example:
            sh bootstrap.sh -P -y -x python2.7 git v2017.7.2
        The above will install python27 and install the git version of salt using the
        python2.7 executable. This only works for git and pip installations.

The Salt Bootstrap script has a wide variety of options that can be passed as
well as several ways of obtaining the bootstrap script itself. Note that the use of ``sudo``
is not needed when running these commands as the ``root`` user.

**NOTE**

The examples below show how to bootstrap Salt directly from GitHub or another Git repository.
Run the script without any parameters to get latest stable Salt packages for your system from
`SaltStack's corporate repository`_. See first example in the `Install using wget`_ section.


Install using curl
~~~~~~~~~~~~~~~~~~

If you want to install a package of a specific release version, from the SaltStack repo:

.. code:: console

  curl -o bootstrap-salt.sh -L https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh -P stable 3004.1

If you want to install a specific release version, based on the Git tags:

.. code:: console

  curl -o bootstrap-salt.sh -L https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh git v3004.1

Using ``curl`` to install latest development version from GitHub:

.. code:: console

  curl -o bootstrap-salt.sh -L https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh git master

To install a specific branch from a Git fork:

.. code:: console

  curl -o bootstrap-salt.sh -L https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh -g https://github.com/myuser/salt.git git mybranch

If all you want is to install a ``salt-master`` using latest Git:

.. code:: console

  curl -o bootstrap-salt.sh -L https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh -M -N git master

If your host has Internet access only via HTTP proxy, from the SaltStack repo:

.. code:: console

  PROXY='http://user:password@myproxy.example.com:3128'
  curl -o bootstrap-salt.sh -L -x "$PROXY" https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh -P -H "$PROXY" stable

If your host has Internet access only via HTTP proxy, installing via Git:

.. code:: console

  PROXY='http://user:password@myproxy.example.com:3128'
  curl -o bootstrap-salt.sh -L -x "$PROXY" https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh -H "$PROXY" git


Install using wget
~~~~~~~~~~~~~~~~~~

Using ``wget`` to install your distribution's stable packages:

.. code:: console

  wget -O bootstrap-salt.sh https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh

Installing a specific version from git using ``wget``:

.. code:: console

  wget -O bootstrap-salt.sh https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh git v3004.1

Installing a specific version package from the SaltStack repo using ``wget``:

.. code:: console

  wget -O bootstrap-salt.sh https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh -P stable 3004.1

**NOTE**

On the above examples we added ``-P`` which will allow PIP packages to be installed if required.
However, the ``-P`` flag is not necessary for Git-based bootstraps.


Install using Python
~~~~~~~~~~~~~~~~~~~~

If you already have Python installed, ``python 2.7``, then it's as easy as:

.. code:: console

  python -m urllib "https://bootstrap.saltproject.io" > bootstrap-salt.sh
  sudo sh bootstrap-salt.sh -P stable 3004.1

With python version 2, the following in-line code should always work:

.. code:: console

  python -c 'import urllib; print urllib.urlopen("https://bootstrap.saltproject.io").read()' > bootstrap-salt.sh
  sudo sh bootstrap-salt.sh git master

With python version 3:

.. code:: console

  python3 -c 'import urllib.request; print(urllib.request.urlopen("https://bootstrap.saltproject.io").read().decode("ascii"))' > bootstrap-salt.sh
  sudo sh bootstrap-salt.sh git v3004.1

Install using fetch
~~~~~~~~~~~~~~~~~~~

On a FreeBSD-based system you usually don't have either of the above binaries available. You **do**
have ``fetch`` available though:

.. code:: console

  fetch -o bootstrap-salt.sh https://bootstrap.saltproject.io
  sudo sh bootstrap-salt.sh

If you have any SSL issues install ``ca_root_nss``:

.. code:: console

  pkg install ca_root_nss

And either copy the certificates to the place where fetch can find them:

.. code:: console

  cp /usr/local/share/certs/ca-root-nss.crt /etc/ssl/cert.pem

Or link them to the right place:

.. code:: console

  ln -s /usr/local/share/certs/ca-root-nss.crt /etc/ssl/cert.pem


Installing via an Insecure One-Liner
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following examples illustrate how to install Salt via a one-liner.

**NOTE**

Warning! These methods do not involve a verification step and assume that the delivered file is
trustworthy.

Any of the examples above which use two lines can be made to run in a single-line
configuration with minor modifications.

Installing the latest stable release of Salt (default):

.. code:: console

  curl -L https://bootstrap.saltproject.io | sudo sh

Using ``wget`` to install your distribution's stable packages:

.. code:: console

  wget -O - https://bootstrap.saltproject.io | sudo sh

Installing a target version package of Salt from the SaltStack repo:

.. code:: console

  curl -L https://bootstrap.saltproject.io | sudo sh -s -- stable 3004.1

Installing the latest master branch of Salt from git:

.. code:: console

  curl -L https://bootstrap.saltproject.io | sudo sh -s -- git master


Install on Windows
~~~~~~~~~~~~~~~~~~

Using ``PowerShell`` to install latest stable version:

.. code:: console

  Invoke-WebRequest -Uri https://winbootstrap.saltproject.io -OutFile C:\Temp\bootstrap-salt.ps1
  Set-ExecutionPolicy -ExecutionPolicy Unrestricted -Scope CurrentUser
  C:\Temp\bootstrap-salt.ps1
  Set-ExecutionPolicy -ExecutionPolicy Undefined -Scope CurrentUser


Using ``cygwin`` to install latest stable version:

.. code:: console

  curl -o bootstrap-salt.ps1 -L https://winbootstrap.saltproject.io
  "/cygdrive/c/WINDOWS/System32/WindowsPowerShell/v1.0/powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "[System.Net.ServicePointManager]::SecurityProtocol = 3072; iex ./bootstrap-salt.ps1"


Supported Operating Systems
---------------------------

The salt-bootstrap script officially supports the distributions outlined in
`Salt's Supported Operating Systems`_ document, except for Solaris and AIX. The operating systems
listed below should reflect this document but may become out of date. If an operating system is
listed below, but is not listed on the official supported operating systems document, the level of
support is "best-effort".

Since Salt is written in Python, the packages available from `SaltStack's corporate repository`_
are CPU architecture independent and could be installed on any hardware supported by Linux kernel.
However, SaltStack does package Salt's binary dependencies only for ``x86_64`` (``amd64``) and
``AArch32`` (``armhf``). The latter is available only for Debian/Raspbian 8 platforms.

It is recommended to use ``git`` bootstrap mode as described above to install Salt on other
architectures, such as ``x86`` (``i386``), ``AArch64`` (``arm64``) or ``ARM EABI`` (``armel``).
You also may need to disable repository configuration and allow ``pip`` installations by providing
``-r`` and ``-P`` options to the bootstrap script, i.e.:

.. code:: console

  sudo sh bootstrap-salt.sh -r -P git master

**NOTE**

Bootstrap may fail to install Salt on the cutting-edge version of distributions with frequent
release cycles such as: Amazon Linux, Fedora, openSUSE Tumbleweed, or Ubuntu non-LTS. Check the
versions from the list below. Also, see the `Unsupported Distro`_ section.


Debian and derivatives
~~~~~~~~~~~~~~~~~~~~~~

- Cumulus Linux 2/3
- Debian GNU/Linux 9/10/11
- Devuan GNU/Linux 1/2
- Kali Linux 1.0 (based on Debian 7)
- Linux Mint Debian Edition 1 (based on Debian 8)
- Raspbian 8 (``armhf`` packages) and 9 (using ``git`` installation mode only)


Red Hat family
~~~~~~~~~~~~~~

- Amazon Linux 2012.3 and later
- Amazon Linux 2
- CentOS 6/7/8
- Cloud Linux 6/7
- Fedora 30/31 (install latest stable from standard repositories)
- Oracle Linux 6/7
- Red Hat Enterprise Linux 6/7/8
- Scientific Linux 6/7


SUSE family
~~~~~~~~~~~

- openSUSE Leap 15 (see note below)
- openSUSE Leap 42.3
- openSUSE Tumbleweed 2015
- SUSE Linux Enterprise Server 11 SP4, 12 SP2

**NOTE:** Leap 15 installs Python 3 Salt packages by default. Salt is packaged by SUSE, and
Leap 15 ships with Python 3. Salt with Python 2 can be installed using the the ``-x`` option
in combination with the ``git`` installation method.

.. code:: console

    sh bootstrap-salt.sh -x python2 git v2018.3.2


Ubuntu and derivatives
~~~~~~~~~~~~~~~~~~~~~~

- KDE neon (based on Ubuntu 18.04)
- Linux Mint 17/18
- Ubuntu 14.04/16.04/18.04 and subsequent non-LTS releases (see below)

Ubuntu Best Effort Support: Non-LTS Releases

ChangeLog
Version TBD (In Progress on the Develop Branch):

Version 2022.05.19
	* Fix situations where the registry key for root_dir does not exist
	on Windows (twangboy) #1828
	* Add Ubuntu 22 support (krionbsd) #1820
	* Add Fedora 36 support (myii) #1818

Version 2022.03.15:
	* Add detection and functions for AlmaLinux and Rocky Linux (myii) #1803
	* Copy configs to correct config dirs (v3004+) (dafyddj) #1798
	* Support Non-LTS Ubuntu 21.04 & 21.10 (blindpirate) #1793
	* Use native repositories for Debian 11 (jpacura) #1615
	* Keep all command-line parameters when UAC is enabled (Simon-TheUser) #1613
	* Add support for Raspbian (Jille) #1612
	* Add openrc to alpine:latest dependencies (krionbsd) #1609
	* Add CentOS 7 base key (bryceml) #1608
	* Fix git master install on alpine 3.12+ (Nascire) #1604
	* Sort help alphabetically (krionbsd) #1601

Version 2021.09.17:
	* Re-add Ubuntu-16 support as it's still supported with 3001 and 3002 (krionbsd) #1594
	* Add oncoming 3004 release (krionbsd) #1593

Version 2021.09.14:
	* Update latest Salt version in README (krionbsd) #1588
	* Make Python 3 the default (bryceml) #1577
	* Don't output info if using the default -x value (bryceml) #1587

Version 2021.08.19:
    * Update README with instructions to use salt-bootstrap custom version (ari) #1580
    * Remove unsupported versions (bryceml) #1578
    * Update FreeBSD default Python version to 3.8 (myii) #1574

Version 2021.06.23:
	* Use salt.list instead of saltstack.list (bryceml) #1563
	* Use fetch_url function for curl (xeacott) #1562
	* Add Ubuntu 21.04 support (krionbsd) #1559
	* Remove python2-futures package for ArchLinux (myii) #1546

Version 2020.03.02:
	* Debian 11 (bullseye/testing) support using Debian 10 packages (jpacura, bryceml) #1514
	* Fix bootstrap on Gentoo (ijansky) #1516 #1518
	* Fix project URLs (bryceml) #1526
	* FreeBSD also needs the _PKI_DIR reset (krionbsd) #1523
	* Fix openSUSE Tumbleweed support. (myii) #1525
	* Fix Debian and Ubuntu version support. (jpacura) #1529
	* Powershell: Implement configure only option (for Vagrant). (dafyddj) #1530
	* Powershell: Handle Vagrant's `grains_config` option. (dafyddj) #1520
	* apt-key is deprecated on Debian 10+ and Ubuntu 20+ (bryceml) #1533
	* Fix python Tornado on FreeBSD. (krionbsd) #1522

Version 2020.10.20:
	* Add support to allow bootstrapping Salt 3002 (s0undt3ch) #1506

Version 2020.10.19:
	* Fix v3000+ with git install on FreeBSD (krionbsd) #1487
	* Update README giving ONE example of WINDOWS bootstrapping. Default to python3 instead of
	python2. (noelmcloughlin) #1496
	* Support git and stable salt-bootstrap on Gentoo. (ijansky) #1500
	* Add support for Linux Mint 20. (taigrr) #1502
	* Adding missing functions for Red Hat 8. (mbochenk) #1489
	* Allow pinning minor 3xxx versions. (max-arnold) #1491

Version 2020.06.23:
	* Fix for Cumulus Linux 4.1 (darylturner) #1474
	* Fix file download exit code, improve error message on failed download (bryceml) #1478
	* Add support for Ubuntu 20.04 (bryceml) #1479
	* Default Ubuntu 20.04 to Py3 since there isn't a Py2 release for it (bryceml) #1480
	* Warn instead of exit when using -R and -x python3 on centos to allow installing python3
	versions of salt on centos from other repos.  Old unsupported releases probably need to
	manually install epel-release before running the bootstrap script for versions before 2018.3.5
	and before 2019.2.1 (bryceml) #1482
	* Start maintaining .sha256 files for the bootstrap script

Version 2020.05.28:
	* Fix Opensuse Tumbleweed (noelmcloughlin) #1441
	* Remove Debian 8 from Python 3 supported list (jay1648) #1448
	* Fixed deps parsing for FreeBSD git installs (cedwards) #1462
	* Added support for the upcoming Sodium release (s0undt3ch) #1466
	* Fixed KDE Neon detection (ender8282) #1466
	* Updated README python 2 and 3 installations (gdebunne) #1466
	* Updated README and CONTRIBUTING to prefer Salt's master branch (DmitryKuzmenko) #1466
	* Fixed IUS repo addresses (oeuftete) #1469
	* Updated tests to assert target Python and Salt versions (s0undt3ch) #1470
	* Fixed Amazon 2 -x support (s0undt3ch) #1470
	* Fix Fedora stable installs for older releases (s0undt3ch) #1470
	* Fix CentOS 8 installations (s0undt3ch) #1470
	* Default Debian 10 to Py3 (s0undt3ch) #1470
	* Fix Arch Linux installs (s0undt3ch) #1470

Version 2020.02.24:
	* Fix SLES 15 install (max298) #1431
	* Fix 3000(Neon) stable install (sblaisot) #1433
	* Fix Amazon Linux Py3 install (s0undt3ch) #1434

Version 2020.02.04:
	* Add support for the Salt Neon(3000) release (s0undt3ch) #1424
	* Drop support for Fedora < 30 (s0undt3ch) #1424
	* Drop support for Debian < 8 (s0undt3ch) #1424

Version 2020.01.29:
	* FreeBSD fixes (cedwards) #1413
	* Support the upcoming Neon release (s0undt3ch) #1420

Version 2020.01.21:
    * FreeBSD fixes (kgbsd) #1376
    * Fix macOS support (s0undt3ch) #1397

Version 2019.11.04:
    * Fix busybox mktemp compatibility (stanzgy) #1369
    * Install debian 10 packages on debian 10 instead of 9 (kiemlicz) #1375
    * move centos to python36, use python specified by -x (Ch3LL,bryceml) #1380
    * Add debian 10 git install support (Ch3LL) #1378

Version 2019.10.03:
    * Fix possible typo with `gnupg-curl` vs `gnupg curl` (zahiar)
    * Install only python3 packges if requested on ubuntu (noelmcloughlin) #1356
    * Fixing debian wheezy (bryceml) #1359
    * Fixed Amazon Linux 2 detection when lsb_release is installed (jars99) #1361
    * Mac OS Support (felippeb) #1363 #1364 #1365 #1366

Version 2019.05.20:
    * Allow stable version selection for amazon linux (puluanau) #1328
    * FreeBSD 12 support (sticky-note) #1329
    * Apt wait for lock modification (ripesensor) #1335
    * Opensuse 15 Git installation fixes (s0undt3ch) #1340 #1341
    * Amazon Linux Image fixes (nshenry03) #1343
    * apt_key_fetch fixes (greut) #1344
    * CentOS Git and PIP installation m2crypto fixes (s0undt3ch) #1347
    * CI process enhancements/fixes (s0undt3ch) #1347

Version 2019.02.27:

.github/workflows/templates/generate.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import datetime

os.chdir(os.path.abspath(os.path.dirname(__file__)))

LINUX_DISTROS = [
    "almalinux-8",
    "amazon-2",
    "arch",
    "centos-7",
    "centos-stream8",
    "debian-10",
    "debian-11",
    "debian-9",
    "fedora-35",
    "fedora-36",
    "gentoo",
    "gentoo-systemd",
    "opensuse-15",
    "opensuse-tumbleweed",
    "oraclelinux-7",
    "oraclelinux-8",
    "rockylinux-8",
    "ubuntu-1804",
    "ubuntu-2004",
    "ubuntu-2204",
]
OSX = WINDOWS = []

STABLE_DISTROS = [
    "almalinux-8",
    "amazon-2",
    "arch",
    "centos-7",
    "centos-stream8",
    "debian-10",
    "debian-11",
    "debian-9",
    "fedora-35",
    "fedora-36",
    "gentoo",
    "gentoo-systemd",
    "opensuse-15",
    "opensuse-tumbleweed",
    "oraclelinux-7",
    "oraclelinux-8",
    "rockylinux-8",
    "ubuntu-1804",
    "ubuntu-2004",
    "ubuntu-2204",
]

BLACKLIST_3003 = [
    "arch",
    "debian-11",
    "fedora-35",
    "fedora-36",
    "gentoo",
    "gentoo-systemd",
    "opensuse-15",
    "opensuse-tumbleweed",
    "rockylinux-8",
    "ubuntu-2204",
]

BLACKLIST_GIT_3003 = [
    "amazon-2",
    "arch",
    "debian-10",
    "debian-11",
    "fedora-35",
    "fedora-36",
    "gentoo",
    "gentoo-systemd",
    "opensuse-15",
    "opensuse-tumbleweed",
    "rockylinux-8",
    "ubuntu-2004",
    "ubuntu-2110",
    "ubuntu-2204",
]

BLACKLIST_3004 = [
    "arch",
    "fedora-35",
    "fedora-36",
    "gentoo",
    "gentoo-systemd",
    "opensuse-15",
    "opensuse-tumbleweed",
]

BLACKLIST_GIT_3004 = [
    "amazon-2",
    "arch",
    "debian-10",
    "debian-11",
    "fedora-35",
    "fedora-36",
    "gentoo",
    "gentoo-systemd",
    "opensuse-15",
    "opensuse-tumbleweed",
    "ubuntu-2004",
    "ubuntu-2110",
    "ubuntu-2204",
]

SALT_BRANCHES = [
    "3003",
    "3004",
    "master",
    "latest",
]

BRANCH_DISPLAY_NAMES = {
    "3003": "v3003",
    "3004": "v3004",
    "master": "Master",
    "latest": "Latest",
}

STABLE_BRANCH_BLACKLIST = []

LATEST_PKG_BLACKLIST = []

DISTRO_DISPLAY_NAMES = {
    "almalinux-8": "AlmaLinux 8",
    "amazon-2": "Amazon 2",
    "arch": "Arch",
    "centos-7": "CentOS 7",
    "centos-stream8": "CentOS Stream 8",
    "debian-10": "Debian 10",
    "debian-11": "Debian 11",
    "debian-9": "Debian 9",
    "fedora-35": "Fedora 35",
    "fedora-36": "Fedora 36",
    "gentoo": "Gentoo",
    "gentoo-systemd": "Gentoo (systemd)",
    "opensuse-15": "Opensuse 15",
    "opensuse-tumbleweed": "Opensuse Tumbleweed",
    "oraclelinux-7": "Oracle Linux 7",
    "oraclelinux-8": "Oracle Linux 8",
    "rockylinux-8": "Rocky Linux 8",
    "ubuntu-1804": "Ubuntu 18.04",
    "ubuntu-2004": "Ubuntu 20.04",
    "ubuntu-2204": "Ubuntu 22.04",
}

TIMEOUT_DEFAULT = 20
TIMEOUT_OVERRIDES = {
    "gentoo": 90,
    "gentoo-systemd": 90,
}
BRANCH_ONLY_OVERRIDES = [
    "gentoo",
    "gentoo-systemd",
]


def generate_test_jobs():
    test_jobs = ""
    branch_only_test_jobs = ""

    for distro in LINUX_DISTROS + OSX + WINDOWS:
        timeout_minutes = (
            TIMEOUT_OVERRIDES[distro]
            if distro in TIMEOUT_OVERRIDES
            else TIMEOUT_DEFAULT
        )
        needs = "    needs: lint"
        if distro in BRANCH_ONLY_OVERRIDES:
            needs = ""
        current_test_jobs = ""

        for branch in SALT_BRANCHES:

            if branch == "latest":
                if distro in LATEST_PKG_BLACKLIST:
                    continue
                if distro in LINUX_DISTROS:
                    template = "linux.yml"
                elif distro in OSX:
                    template = "osx.yml"
                elif distro in WINDOWS:
                    template = "windows.yml"
                else:
                    print("Don't know how to handle {}".format(distro))

                with open(template) as rfh:
                    current_test_jobs += "\n{}\n".format(
                        rfh.read()
                        .replace(
                            "{python_version}-{bootstrap_type}-{branch}-{distro}",
                            "{branch}-{distro}",
                        )
                        .format(
                            distro=distro,
                            branch=branch,
                            display_name="{} Latest packaged release".format(
                                DISTRO_DISPLAY_NAMES[distro],
                            ),
                            timeout_minutes=timeout_minutes,
                            needs=needs,
                        )
                    )
                continue

            for python_version in ("py3",):

                for bootstrap_type in ("stable", "git"):
                    if bootstrap_type == "stable":
                        if branch == "master":
                            # For the master branch there's no stable build
                            continue
                        if distro not in STABLE_DISTROS:
                            continue

                        if branch in STABLE_BRANCH_BLACKLIST:
                            continue

                        if distro.startswith("fedora") and branch != "latest":
                            # Fedora does not keep old builds around
                            continue

                    BLACKLIST = {
                        "3003": BLACKLIST_3003,
                        "3004": BLACKLIST_3004,
                    }
                    if bootstrap_type == "git":
                        BLACKLIST = {
                            "3003": BLACKLIST_GIT_3003,
                            "3004": BLACKLIST_GIT_3004,
                        }

                        # .0 versions are a virtual version for pinning to the first point release of a major release, such as 3003, there is no git version.
                        if branch.endswith("-0"):
                            continue

                    if (
                        branch in ("3003", "3004")
                        and distro in BLACKLIST[branch]
                    ):
                        continue

                    if distro in LINUX_DISTROS:
                        template = "linux.yml"
                    elif distro in OSX:
                        template = "osx.yml"
                    elif distro in WINDOWS:
                        template = "windows.yml"
                    else:
                        print("Don't know how to handle {}".format(distro))

                    with open(template) as rfh:
                        current_test_jobs += "\n{}\n".format(
                            rfh.read().format(
                                distro=distro,
                                branch=branch,
                                python_version=python_version,
                                bootstrap_type=bootstrap_type,
                                display_name="{} {} {} {}".format(
                                    DISTRO_DISPLAY_NAMES[distro],
                                    BRANCH_DISPLAY_NAMES[branch],
                                    python_version.capitalize(),
                                    bootstrap_type.capitalize(),
                                ),
                                timeout_minutes=timeout_minutes,
                                needs=needs,
                            )
                        )
        if distro in BRANCH_ONLY_OVERRIDES:
            branch_only_test_jobs += current_test_jobs
        else:
            test_jobs += current_test_jobs

    with open("lint.yml") as rfh:
        lint_job = "\n{}\n".format(rfh.read())

    with open("pre-commit.yml") as rfh:
        pre_commit_job = "\n{}\n".format(rfh.read())

    with open("../main.yml", "w") as wfh:
        with open("main.yml") as rfh:
            wfh.write(
                "{}\n".format(
                    rfh.read()
                    .format(
                        jobs="{pre_commit}{lint}{test}".format(
                            lint=lint_job,
                            test=test_jobs,
                            pre_commit=pre_commit_job,
                        ),
                        on="push, pull_request",
                        name="Testing",
                    )
                    .strip()
                )
            )

    with open("../main-branch-only.yml", "w") as wfh:
        with open("main.yml") as rfh:
            wfh.write(
                "{}\n".format(
                    rfh.read()
                    .format(
                        jobs="{test}".format(
                            test=branch_only_test_jobs,
                        ),
                        on="push",
                        name="Branch Testing",
                    )
                    .strip()
                )
            )


if __name__ == "__main__":
    generate_test_jobs()
